src/main/java/me/itzg/tsdbcassandra/services/DownsampleProcessor.java

```diff
@@ -0,0 +1,216 @@
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+import reactor.core.scheduler.Scheduler;
+import reactor.util.function.Tuple2;
+
+@Service
+@Slf4j
+public class DownsampleProcessor {
+
+  private final Environment env;
+  private final DownsampleProperties downsampleProperties;
+  private final DownsampleTrackingService downsampleTrackingService;
+  private final Scheduler scheduler;
+  private final SeriesSetService seriesSetService;
+  private final QueryService queryService;
+  private final DataWriteService dataWriteService;
+  private final MetadataService metadataService;
+  private Disposable scheduled;
+
+  @Autowired
+  public DownsampleProcessor(Environment env,
+                             DownsampleProperties downsampleProperties,
+                             DownsampleTrackingService downsampleTrackingService,
+                             @Qualifier("downsampleScheduler") Scheduler downsampleScheduler,
+                             SeriesSetService seriesSetService,
+                             QueryService queryService,
+                             DataWriteService dataWriteService,
+                             MetadataService metadataService) {
+    this.env = env;
+    this.downsampleProperties = downsampleProperties;
+    this.downsampleTrackingService = downsampleTrackingService;
+    scheduler = downsampleScheduler;
+    this.seriesSetService = seriesSetService;
+    this.queryService = queryService;
+    this.dataWriteService = dataWriteService;
+    this.metadataService = metadataService;
+  }
+
+  @PostConstruct
+  public void setupSchedulers() {
+    if (env.acceptsProfiles(Profiles.of("test"))) {
+      log.warn("Downsample scheduling disabled during testing");
+      return;
+    }
+
+    if (downsampleProperties.getPartitionsToProcess() == null ||
+        downsampleProperties.getPartitionsToProcess().isEmpty()) {
+      // just info level since this is the normal way to disable downsampling
+      log.info("Downsample processing is disabled due to no partitions to process");
+      return;
+    }
+
+    if (downsampleProperties.getGranularities() == null ||
+        downsampleProperties.getGranularities().isEmpty()) {
+      throw new IllegalStateException("Downsample partitions to process are configured, but not granularities");
+    }
+
+    scheduled = Flux.interval(downsampleProperties.getDownsampleProcessPeriod(), scheduler)
+        .flatMap(tick -> Flux.fromIterable(downsampleProperties.getPartitionsToProcess()))
+        .flatMap(this::processPartition)
+        .subscribe();
+
+    log.debug("Downsample processing is scheduled");
+  }
+
+  @PreDestroy
+  public void stop() {
+    if (scheduled != null) {
+      scheduled.dispose();
+    }
+  }
+
+  private Flux<?> processPartition(int partition) {
+    log.trace("Downsampling partition {}", partition);
+
+    return downsampleTrackingService
+        .retrieveReadyOnes(partition)
+        .publishOn(scheduler)
```

Comment:
so it looks like processPartition() is being called from the Flux.interval() that is being invoked with the scheduler.  If I'm understanding correctly, that would make this call to publishOn() redundant?

Reply:
Yes, to the first part;  however, stay tuned, because I'm now getting rid of `Flux.interval` in the current branch since the error propagation was terminating the recurring task execution I wanted there.

As for `publishOn` I'm not feeling very confident about the uses of those, but I added them because the cassandra calls must be doing their own `publishOn` operators because the logging shows it hopping over to one of their "s0-io" threads like:

```
2020-09-18 10:53:20.785 DEBUG 57857 --- [   downsample-2] m.i.t.services.DownsampleProcessor       : Processing downsample set PendingDownsampleSet(partition=0, timeSlot=2020-09-18T15:42:00Z, tenant=geof0549, seriesSet=processes_sleeping,host=MS90HCG8WL, lastTouch=2020-09-18T15:42:31.835Z)
2020-09-18 10:53:20.785 DEBUG 57857 --- [        s0-io-3] m.i.t.services.DownsampleProcessor       : Completed downsampling of PendingDownsampleSet(partition=0, timeSlot=2020-09-18T15:42:00Z, tenant=geof0549, seriesSet=swap_in,host=MS90HCG8WL, lastTouch=2020-09-18T15:42:31.707Z)
```

With the metrics gathering added to the new branch, I found in local testing that the measured durations were not noticeably impacted by including or removing those `publishOn` operators. So now I'm considering removing them since they're working against the reactive cassandra threading and I now worry may actually introduce more thread-context switching overhead. That would be a shame since I think event-driven/reactive programming strives to avoid that kind of overhead.

Reply:
...FYI the bit in brackets `[   downsample-2]` is the thread name, where, as you probably saw, the scheduler I am using was configured with the thread name prefix "downsample".